Templates/DefaultProject/Template/cmake/VersionUtils.cmake

```diff
@@ -0,0 +1,116 @@
+# {BEGIN_LICENSE}
```

Comment:
The template.json hasn't been updated to copy over this file when creating a project..

The MinimalProject template also needs this update.

Reply:
Furthermore do not, that updating existing projects to use newer engine_finder cmake scripts is not easy to do, so we should endeavor to have as few files and as few changes as necessary to locate and validate the engine.

Once users make a project using these templates, it will much harder to update the engine finder script, so we should account for any potential issues where this file would need to be updated in the future.

Reply:
Good catch.  I created VersionUtils as a library that I thought might be useful to anyone making their own custom `EngineFinder.cmake`.  I'm also leaning towards adding the new `EngineFinder.cmake` changes as a new file like you mentioned - called `EngineFinderV2.cmake` and then just modify the `CMakeList.txt` to use it as the new default.  That way when we update user projects (code to be written)  it would leave the original `EngineFinder.cmake` and they could easily switch back to it if they wanted to.

Reply:
Hmmm, I suppose what you're getting at is user's might modify the VersionUtils.cmake file and make it harder if we wanted to update it...

Reply:
Well if the user modifies the VersionUtils.cmake, then they are in advanced mode.
I more thinking of if we find a bug in VersisonUtils.cmake or EngineFinder.cmake it harders to push the changes out to existing projects.

By trying to keep it the logic as lightweight as possible, we can reduce the chances that a user needs to upgrade a project.

The point I was trying to get across is that it is a lot easier for a user to create a new project with the latest EngineFinder logic, then it is for them to upgrade an existing project to use the latest EngineFinder.cmake.

So we should try to avoid the need for the user to upgrade the EngineFinder.cmake

Reply:
I was considering putting the compatibility check logic inside the engine to keep the engine finder in the project more simple.  e.g. for every engine path:

1. check if a `<engine root>/cmake/ProjectCompatibility.cmake` exists , and if so, include it.  Inside that file the engine would look at vars like `O3DE_ENGINE_NAME_TO_USE` and friends and would set some variable called `O3DE_COMPATIBLE_ENGINE_VERSION`  and `O3DE_COMPATIBLE_ENGINE_PATH` etc. - but only if that `O3DE_COMPATIBLE_ENGINE_PATH`variable wasn't set, or if the engine had a higher compatible version.   
2. if the `<engine root>/cmake/ProjectCompatibility.cmake` wasn't found in the engine path then assume compatible if no `O3DE_COMPATIBLE_ENGINE_PATH` has already been set by an engine that did check compatibility.
![image](https://user-images.githubusercontent.com/26804013/216734792-c9389324-1ff6-4138-aaad-a6bbb90c34e6.png)

Reply:
That could work.
Technically the only code the project needs access to is the logic to lookup the list of paths of engines in the `o3de_manifest.json`.
Once a set of candidate engine paths are available, the engine can do the rest of the lifting.

The minimum set of operations the EngineFinder.cmake needs to support each of the following three use cases are as follows
1. The user specifies the `-DCMAKE_MODULE_PATH` through the CMake command line directly, well that is the complete set of "candidate engine paths" to check in this scenario
2. The `<project-root>/user/project.json` has an `engine_path` key, that is added to the set of "candidate engine paths"
3. The `<project-root>/user/project.json` doesn't have a `engine_path` key, in that case the project only needs to query the o3de_manifest.json for the set of engines from the "engines" array and append it to it's "candidate engine paths"

Afterwards the only other operations is looping over the "candidate engine paths" and including `<engine root>/cmake/ProjectCompatibility.cmake` for each engine.